File: Sources/vfx/export_by_markers.py
# -*- coding: utf-8 -*-
import sys
import os

# Раскомментировано для macOS:
sys.path.insert(0, "/Library/Application Support/Blackmagic Design/DaVinci Resolve/Developer/Scripting/Modules")

import DaVinciResolveScript as dvr

resolve = dvr.scriptapp("Resolve")
pm = resolve.GetProjectManager()
project = pm.GetCurrentProject()
timeline = project.GetCurrentTimeline()

if not timeline:
    print("Нет активного таймлайна!")
    sys.exit(1)

markers = timeline.GetMarkers()

if not markers:
    print("Маркеры не найдены!")
    sys.exit(1)

# Папка для экспорта — измените под себя
output_path = os.path.expanduser("~/Desktop/Export")
os.makedirs(output_path, exist_ok=True)

fps = float(timeline.GetSetting("timelineFrameRate"))

for frame_id, marker in markers.items():
    name = marker.get("name").strip()
    duration = marker.get("duration", 1)

    if not name:
        print(f"  Пропущен маркер на фрейме {frame_id} — нет имени")
        continue

    start = frame_id
    end = frame_id + duration - 1

    print(f"Добавляю: '{name}' ({start} → {end})")

    project.SetRenderSettings({
        "SelectAllFrames": False,
        "MarkIn": start,
        "MarkOut": end,
        "CustomName": name,
        "TargetDir": output_path,
    })

    project.AddRenderJob()

print(f"\nДобавлено задач: {len(markers)}. Запускаю рендер...")
project.StartRendering()
print("Готово!")

File: Sources/vfx/export_by_markers.lua
-- DaVinci Resolve Script: Export by Markers
-- Автоматический пакетный экспорт клипов по маркерам

local resolve = Resolve()
local pm = resolve:GetProjectManager()
local project = pm:GetCurrentProject()

if not project then
    print("Ошибка: проект не открыт")
    return
end

local timeline = project:GetCurrentTimeline()

if not timeline then
    print("Ошибка: нет активного таймлайна")
    return
end

local markers = timeline:GetMarkers()

if not markers or next(markers) == nil then
    print("Ошибка: маркеры не найдены")
    return
end

-- Папка для экспорта
local output_path = os.getenv("HOME") .. "/Desktop/Export"

-- Создание директории
os.execute("mkdir -p '" .. output_path .. "'")

-- Получение FPS
local fps = tonumber(timeline:GetSetting("timelineFrameRate"))

local marker_count = 0

for frame_id, marker in pairs(markers) do
    local name = marker["name"]

    if name and name ~= "" then
        name = name:match("^%s*(.-)%s*$") -- trim

        if name ~= "" then
            local duration = marker["duration"] or 1
            local start_frame = frame_id
            local end_frame = frame_id + duration - 1

            print(string.format("Добавляю: '%s' (фреймы %d → %d)", name, start_frame, end_frame))

            -- Установка параметров рендера
            project:SetRenderSettings({
                SelectAllFrames = false,
                MarkIn = start_frame,
                MarkOut = end_frame,
                CustomName = name,
                TargetDir = output_path,
            })

            -- Добавление задачи рендера
            project:AddRenderJob()
            marker_count = marker_count + 1
        end
    end
end

if marker_count > 0 then
    print(string.format("\nДобавлено задач: %d", marker_count))
    print("Запуск рендера...")
    project:StartRendering()
    print("Готово!")
else
    print("Ошибка: не найдены маркеры с названиями")
end
